### ISSUE TITLE
TypeScript compiler hangs case

### ISSUE THREAD

**Issue by falsandtru** [Microsoft/TypeScript#4604]
Hangs on https://github.com/Microsoft/TypeScript/commit/06e25377ffdb0f89550f1a418d3e769fdfc3de7e

Reproduction

1. checkout https://github.com/falsandtru/one-liner-stream/commit/7804f9125f1b5040f3a8c85b599dc57df125c578
2. checkout and build https://github.com/Microsoft/TypeScript/commit/06e25377ffdb0f89550f1a418d3e769fdfc3de7e
3. `$ cd one-liner-stream `
4.  `$ node ../TypeScript/built/local/tsc.js src/ts/*.ts typings/*.ts --module commonjs --noEmit`

Node 0.12.7

**Comment by falsandtru** [comment]
@mhegazy please.

**Comment by DanielRosenwasser** [comment]
@falsandtru can you please provide us with a simple `tsconfig.json` to compile your project? Ideally a more minimal reproduction of the issue. Otherwise it's difficult for me to verify that this is indeed a problem with the TypeScript and figure out what could be causing it.

**Comment by falsandtru** [comment]
I do not use `tsconfig.json` to this project. I'll try to minimize. I want to make sure in advance, were you able to reproduce this hangs? Thanks.

**Comment by falsandtru** [comment]
Please compile this code.

run `$ node built/local/tsc.js export.ts`

```ts
// export.ts
export class ReadableLiner<T> {
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T>) => TransformLiner<U, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T>) => TransformLiner<T, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, U>) => TransformLiner<U, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T|U>) => TransformLiner<T|U, any>|PassThroughLiner<U>) {
    return composer(this);
  }
}

export class TransformLiner<T, U> {
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T>) => TransformLiner<U, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T>) => TransformLiner<T, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, U>) => TransformLiner<U, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T|U>) => TransformLiner<T|U, any>|PassThroughLiner<U>) {
    return composer(this);
  }
}

export class PassThroughLiner<T> {
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T>) => TransformLiner<U, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T>) => TransformLiner<T, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, U>) => TransformLiner<U, U>|PassThroughLiner<U>): TransformLiner<U, any>|PassThroughLiner<U>
  compose<U>(composer: (stream: ReadableLiner<T>|TransformLiner<T, T|U>) => TransformLiner<T|U, any>|PassThroughLiner<U>) {
    return composer(this);
  }
}

new ReadableLiner<number>()
  .compose(stream => stream);

```

**Comment by falsandtru** [comment]
More info:
My original code compile finished without errors after a few hours, but compiled code is empty. **TS compiler must be show the error in this case because compile is failed.**

**Comment by DanielRosenwasser** [comment]
We're going to need to investigate; in the mean time, this will avoid the issue you're running into:

```TypeScript
interface Liner<T> {
    compose<U>(composer: (stream: ReadableLiner<T> | TransformLiner<T, T>) => TransformLiner<U, U> | PassThroughLiner<U>): TransformLiner<U, any> | PassThroughLiner<U>
    compose<U>(composer: (stream: ReadableLiner<T> | TransformLiner<T, T>) => TransformLiner<T, U> | PassThroughLiner<U>): TransformLiner<U, any> | PassThroughLiner<U>
    compose<U>(composer: (stream: ReadableLiner<T> | TransformLiner<T, U>) => TransformLiner<U, U> | PassThroughLiner<U>): TransformLiner<U, any> | PassThroughLiner<U>
}

export class ReadableLiner<T> implements Liner<T> {
    compose<U>(composer: (stream: ReadableLiner<T> | TransformLiner<T, T | U>) => TransformLiner<T | U, any> | PassThroughLiner<U>) {
        return composer(this);
    }
}

export class TransformLiner<T, Z> implements Liner<T> {
    compose<U>(composer: (stream: ReadableLiner<T> | TransformLiner<T, T | U>) => TransformLiner<T | U, any> | PassThroughLiner<U>) {
        return composer(this);
    }
}

export class PassThroughLiner<T> implements Liner<T> {
    compose<U>(composer: (stream: ReadableLiner<T> | TransformLiner<T, T | U>) => TransformLiner<T | U, any> | PassThroughLiner<U>) {
        return composer(this);
    }
}

var x = new ReadableLiner().compose(stream => stream)

new ReadableLiner<number>()
    .compose(stream => stream);
```

**Comment by falsandtru** [comment]
I confirmed, thank you. Another question. Are you accepted the another bug about error messages? I must be open the new issue?

**Comment by ahejlsberg** [comment]
Fixed in #5738 and #5942.
